### PR TITLE
feat: Signature Verification Digest Update of Container Instances

### PR DESCRIPTION
**Pull request by sfiorani**
This PR combines the different features provided by the Container Instance and the new Signature Verification service: indeed, if a Container Instance doesn't provide the Enforcement Digest option (added in #5197 ) but a Trust Anchor is given, Kura will try to calcuate the image digest through the Signature Verification service.

 If no instances of the latter service are available or the verification fails, the Container Instance will continue to provide an empty Enforcement Digest option.

This mechanism is part of the Container Authenticity feature introduced in Kura. Docs will be provided in #5205 

**Related Issue:**

**Description of the solution adopted:**

**Screenshots:**

**Manual Tests:**

**Any side note on the changes made:**

### CODE REVIEW

File: kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java [hunk 1]
```diff
@@ -49,8 +50,9 @@ public class ContainerInstance implements ConfigurableComponent, ContainerOrches
 
     private ContainerOrchestrationService containerOrchestrationService;
     private Set<ContainerSignatureValidationService> availableContainerSignatureValidationService = new HashSet<>();
-
+    private ConfigurationService configurationService;
     private State state = new Disabled(new ContainerInstanceOptions(Collections.emptyMap()));
+    private ContainerInstanceOptions actualOptions = null;
```
Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: Mmmmmmh I think you fell for a classic "false friend" blunder here. You probably wanted to say *current* options, didn't you?

```suggestion
    private ContainerInstanceOptions currentOptions = null;
```

File: kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java [hunk 2]
```diff
@@ -90,25 +100,35 @@ public void updated(Map<String, Object> properties) {
...
-                        containerSignatureValidated.isSignatureValid() ? "OK" : "FAIL");
-            } else {
-                logger.info("No trust anchor available. Signature validation skipped.");
-            }
 
-            if (newProps.isEnabled()) {
-                this.containerOrchestrationService.registerListener(this);
-            } else {
-                this.containerOrchestrationService.unregisterListener(this);
+            if (this.actualOptions == null || !this.actualOptions.equals(new ContainerInstanceOptions(properties))) {
```
Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: As always I would suggest you to use early returns here for improved readability

```suggestion
            if (this.actualOptions != null && this.actualOptions.equals(new ContainerInstanceOptions(properties))) {
                return;
            }
```

Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: Additionally, instead of building the `ContainerInstanceOptions` object so many times (which is a costly operation) I would store it in a variable before checking if it equals the `currentOptions`. Doing so will also improve readability:

```suggestion
            ContainerInstanceOptions newOptions = new ContainerInstanceOptions(properties);
            if (this.currentOptions != null && this.currentOptions.equals(newOptions)) {
```

File: kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java [hunk 3]
```diff
@@ -425,4 +445,42 @@ public State onDisabled() {
 
     }
 
-}
\ No newline at end of file
+    private void getDigestFromSignatureVerification(Map<String, Object> properties) {
```
Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: This could be personal taste but I think this method is hiding quite a lot of things... for starters it's not actually only `get`ting the digest, it is also updating the properties **and** the snapshot.

Personally I would make it return the digest and I would perform the properties/snapshot update in the caller method. This would render this method useless though...

At a minimum we need to update the name. Let's discuss this offline though.

File: kura/test/org.eclipse.kura.container.orchestration.provider.test/src/test/java/org/eclipse/kura/container/orchestration/provider/ContainerOrchestrationServiceImplTest.java [hunk 4]
```diff
@@ -663,6 +664,11 @@ private void whenStopContainer() throws KuraException {
         this.dockerService.stopContainer(this.containerId);
     }
 
+    private void whenDeleteContainer() throws KuraException {
+        // deleteContainer
+        this.dockerService.deleteContainer(this.containerId);
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.container.orchestration.provider.test/src/test/java/org/eclipse/kura/container/orchestration/provider/ContainerOrchestrationServiceImplTest.java]: Is this comment really needed though? 😄 

```suggestion
        this.dockerService.deleteContainer(this.containerId);
```

File: kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java [hunk 5]
```diff
@@ -502,7 +540,10 @@ private void givenContainerSignatureValidationServiceReturningSuccessFor(String
         // Generate random sha256 string
         String sha256 = "sha256:" + Long.toHexString(Double.doubleToLongBits(Math.random()));
         when(this.mockContainerSignatureValidationService.verify(eq(imageName), eq(imageTag), any(String.class),
-                any(Boolean.class))).thenReturn(new ValidationResult(true, sha256));
+                any(Boolean.class))).thenAnswer(answer -> {
+                    this.properties.put("enforcement.digest", sha256);
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java]: I don't like this at all... why do you need it?

File: kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java [hunk 6]
```diff
@@ -609,6 +656,12 @@ private void thenAuthenticatedVerifySignatureWasCalledFor(String imageName, Stri
                 verifyTlog, passwordRegistryCredentials);
     }
 
+    private void thenSignatureVerificationVerifyWasNeverCalledFor(String imageName, String imageTag, String trustAnchor,
+            boolean verifyTlog, PasswordRegistryCredentials passwordRegistryCredentials) throws KuraException {
+        verify(this.mockContainerSignatureValidationService, never()).verify(imageName, imageTag, trustAnchor,
+                verifyTlog, passwordRegistryCredentials);
+    }
+
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java]: Why did you need this when there's already: https://github.com/eclipse/kura/blob/3401ffe18656115a03500c2a00f0689a83554f6d/kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java#L640

File: kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java [hunk 7]
```diff
@@ -365,6 +372,7 @@ public void signatureValidationWorks() throws KuraException, InterruptedExceptio
         givenPropertiesWith(CONTAINER_IMAGE_TAG, "latest");
         givenPropertiesWith(CONTAINER_TRUST_ANCHOR, "aRealTrustAnchor ;)");
         givenPropertiesWith(CONTAINER_VERIFY_TLOG, true);
+        givenPropertiesWith(CONTAINER_ENFORCEMENT_DIGEST, "");
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java]: If this is needed here I would suspect it was needed for all the other signature-related tests... otherwise they would fail for the wrong reasons... the same goes for the `givenMockedConfigurationService();`

File: kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java [hunk 8]
```diff
@@ -57,6 +62,7 @@ public class ContainerInstanceTest {
     private static final String CONTAINER_VERIFY_TLOG = "container.signature.verify.transparency.log";
     private static final String CONTAINER_REGISTRY_USERNAME = "registry.username";
     private static final String CONTAINER_REGISTRY_PASSWORD = "registry.password";
+    private static final String CONTAINER_ENFORCEMENT_DIGEST = "enforcement.digest";
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java:65]: I noticed this only now but... for consistency, since it's releated to the other parameters, would it make sense to call it:

```suggestion
    private static final String CONTAINER_ENFORCEMENT_DIGEST = "container.signature.enforcement.digest";
```

Obviously such rename would require a separate PR. I'm just throwing this here in the meantime...

**Review by mattdibi — CHANGES_REQUESTED** [review]
Looks good. Some minor things here and there.

File: kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java (hunk 1 above)
Comment (sfiorani) [reply kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: I did it **110%**

**Review by mattdibi — APPROVED** [review]

**Review by mattdibi — CHANGES_REQUESTED** [review]
I just noticed that the "Container Image Enforcement Digest" box is at the bottom of the Container Instance configuration UI. I think its proper place would be under the "Verify In transparency log" radio button as per our initial sketch... let's discuss this tomorrow

image

File: kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java [hunk 9]
```diff
@@ -90,25 +100,51 @@ public void updated(Map<String, Object> properties) {
...
+                logger.info(
+                        "Container configuration doesn't include enforcement digest. Validating with Container Signature Validation service");
+
+                if (this.currentOptions.getSignatureTrustAnchor().isPresent()) {
+
+                    ValidationResult containerSignatureValidated = validateContainerImageSignature(this.currentOptions);
+
+                    logger.info("Container signature validation result for {}@{}({}) - {}",
+                            this.currentOptions.getContainerImage(),
+                            this.currentOptions.getEnforcementDigest().orElse("?"),
```
Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/src/main/java/org/eclipse/kura/container/provider/ContainerInstance.java]: If we want to display the resulting digest we should log `containerSignatureValidated`... not the current options.

```suggestion
                            containerSignatureValidated.imageDigest().orElse("?"),
```

**Review by mattdibi — APPROVED** [review]

**Comment by mattdibi** [comment]
The build succeeds locally but we are facing an issue in the CI connection to Sonar. An issue has been raised few weeks ago to the Eclipse Team: https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/4422

Merging as-is for now.

File: kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java (hunk 8 above)
Comment (sfiorani) [reply kura/test/org.eclipse.kura.container.provider.test/src/test/java/org/eclipse/kura/container/provider/ContainerInstanceTest.java:65]: Done in #5208
